src/librbd/api/Mirror.cc

```diff
@@ -3163,102 +3165,104 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
     req->send();
     int r = cond.wait();
     if (r < 0) {
-      lderr(cct) << "failed to list snapshots in the group " << group_name
+      lderr(cct) << "failed to list snapshots in the group " << group_id
                  << " :" << cpp_strerror(r) << dendl;
       return r;
     }
     auto snap =  snaps.rbegin();
     for (; snap != snaps.rend(); ++snap) {
-      if (snap->state == cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE) {
+      auto snap_type = cls::rbd::get_group_snap_namespace_type(
+          snap->snapshot_namespace);
+      if (snap_type != cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_MIRROR ||
+          snap->state == cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE) {
         continue;
       }
       break;
     }
 
-    if (snap != snaps.rend()) {
-      // Check for group membership match
-      std::vector<cls::rbd::GroupImageSpec> rollback_images;
-      for (auto& it : snap->snaps) {
-        rollback_images.emplace_back(it.image_id, it.pool);
-      }
+    if (snap == snaps.rend()) {
+      lderr(cct) << "failed to rollback, no mirror group snapshot available"
```

Comment:
s/no mirror group snapshot/no complete mirror group snapshot/ ?

Reply:
This will leave a window for the reader if there is an incomplete group snap available here :-)
Hence I will stick to the existing as we cannot conclude it in the log if we have an incomplete snap.